src/error.rs

```diff
@@ -89,21 +122,19 @@ impl Error {
     /// the case that the underlying error is of a different type, the
     /// original Error is returned.
     pub fn downcast<T: Fail>(self) -> Result<T, Error> {
-        let ret: Option<T> = self.downcast_ref().map(|fail| {
-            unsafe {
-                // drop the backtrace
-                let _ = ptr::read(&self.inner.backtrace as *const Backtrace);
-                // read out the fail type
-                ptr::read(fail as *const T)
-            }
-        });
-        match ret {
-            Some(ret) => {
-                // forget self (backtrace is dropped, failure is moved
-                mem::forget(self);
-                Ok(ret)
-            }
-            _       => Err(self)
+        if self.downcast_ref::<T>().is_some() {
```

Comment:
I would like more comments on how this rewrite of the unsafe code works and is correct. To me it looks like this is wrong, but I'm not confident I'm right about that.

In particular, if there is a backtrace, you don't seem to ever actually drop the backtrace here. This is what the `ptr::read` stuff was about.

Reply:
Sure, I'll add a comment. The drop is occurring implicitly since `Box::from_raw(ptr as *mut WithBacktrace<T>)` creates a `Box<WithBacktrace<T>>`. The `failure` field is moved out, and the rest of the box (including the remaining backtrace field) is dropped.